File: benchmarks/livermore/k09_integrate.c
/* K9 — Integrate Predictors (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION PX(101,25), N=101
   PX(I,1) = DM28*PX(I,13) + DM27*PX(I,12) + DM26*PX(I,11)
           + DM25*PX(I,10) + DM24*PX(I,9)  + DM23*PX(I,8)
           + DM22*PX(I,7)  + C0*(PX(I,5) + PX(I,6)) + PX(I,3)
   Column-major: PX(I,J) at flat (J-1)*101 + I (1-based) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     101
#define NCOL  25
#define NREPS 10000

/* 1-based Fortran column-major: PX(i,j) with dims (101,25) */
#define PX(i,j) px[((j)-1)*101 + (i)]

int main(void) {
    /* Max flat index: (25-1)*101 + 101 = 2525, need [2526] */
    double px[2526];
    double spacer[40]; signel(spacer+1, 39);
    double dm28 = spacer[22], dm27 = spacer[21], dm26 = spacer[20];
    double dm25 = spacer[19], dm24 = spacer[18], dm23 = spacer[17];
    double dm22 = spacer[16], c0 = spacer[12];

    /* Initialise: fill 101*25=2525 elements starting at index 1 */
    signel(&px[1], 101 * 25);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long l = 0; l < NREPS; l++) {
        for (long i = 1; i <= N; i++) {
            PX(i,1) = dm28*PX(i,13) + dm27*PX(i,12)
                     + dm26*PX(i,11) + dm25*PX(i,10)
                     + dm24*PX(i,9)  + dm23*PX(i,8)
                     + dm22*PX(i,7)  + c0*(PX(i,5) + PX(i,6))
                     + PX(i,3);
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("px(51,1) = %f\n", PX(51,1));
    return 0;
}

File: benchmarks/livermore/k09_integrate.f
C     Kernel 9 -- Integrate predictors
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K09
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION PX(101,25)
      DIMENSION SPACER(39)
C
C     Initialize spacer for constants
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      SCALED = 0.1D0
      DO 5 K = 1, 39
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          SPACER(K) = (BUZZ - FIZZ) * SCALED
    5 CONTINUE
C     C indices (0-based): 15,16,17,18,19,20,21,11
C     Fortran (1-based): 16,17,18,19,20,21,22,12
      DM22 = SPACER(16)
      DM23 = SPACER(17)
      DM24 = SPACER(18)
      DM25 = SPACER(19)
      DM26 = SPACER(20)
      DM27 = SPACER(21)
      DM28 = SPACER(22)
      C0   = SPACER(12)
C
C     Initialize PX using SIGNEL pattern
C     C: px[101][25], stored row-major
C     Fortran: PX(101,25), stored column-major
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 12 J = 1, 25
          DO 10 I = 1, 101
              BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
              FUZZ = -FUZZ
              PX(I,J) = (BUZZ - FIZZ) * SCALED
   10     CONTINUE
   12 CONTINUE
C
C     Kernel loop
C     C uses 0-based column indices; Fortran uses 1-based
C     C: px[i][0] = ... px[i][12] ... px[i][2]
C     F: PX(I,1) = ... PX(I,13) ... PX(I,3)
C
      N = 101
      DO 200 REP = 1, 10000
          DO 90 I = 1, N
              PX(I,1) = DM28*PX(I,13) + DM27*PX(I,12)
     &                + DM26*PX(I,11) + DM25*PX(I,10)
     &                + DM24*PX(I,9)  + DM23*PX(I,8)
     &                + DM22*PX(I,7)  + C0*(PX(I,5) + PX(I,6))
     &                + PX(I,3)
   90     CONTINUE
  200 CONTINUE
C
      WRITE(*,900) PX(51,1)
  900 FORMAT('K9 integrate pred: px(51,1) = ', E25.15)
      END
